Reorg + JIT Document building

Gregoor opened the pull request:
The long awaited sequel to #762 

This time with a new feature: It fixes #832, fixes #824 

Fixes #976

Alrighty, you two. The framework is in place. Here's my idea for the way forward (ofc feel free to make other suggestions):

I'd work on (re)adding
- [x] sitemaps
- [x] live samples
  - [x] live sample building
- [x] make the functional tests pass
- [x] CRUD
- [x] Document watcher
- [x] updating documentation
- [x] support additional content root (aka archive)
- [x] build search index json
- [x] memoize and make it fast
- [x] `yarn build` should exit non-zero if an error happens
- [x] [search should wait to response on XHR requests](https://github.com/mdn/yari/issues/976)
___

@peterbe & @escattone: These are some other bigger remaining TODOs, I don't have a clear grasp on who'd be best for which one. But if you want to tell me which one(s) you want to take on, I'd create some sprint task issues to satisfy the sprint gods.

## ~~translations~~
~~As we three talked about the current lookup model doesn't work for reductio (`translation_of` living in the translated document, hence an O(1) lookup not being possible). It sounded like we were all vaguely in favor of having something like a `translations.json` in the locale root (just like `wiki_history.json`). But I don't want to speak for anyone else if there's still disagreement here.
Anyway, if there isn't disagreement. This would require a change to the importer (which now lives in its own workspace)~~

**Update:** Ryan and I talked about this yesterday, he made the good point that we don't know how to move forward with translations in general, so we shouldn't block on this. We know how to solve this in a reductio world but the effort might not be worth it at the moment.

## flaws
I think I broke some things there and I haven't backported (err frontported?) fixable flaws yet. Also happy to take this on as I broke it, but if you feel inclined to, that's also great.

## everything KS
I touched a lot of that code and haven't exhaustively checked the results yet. After a general review is done, I'd wanna compare output of the builder on `master` with this branch's output.

## things that were lost
Not on my todo list are stumptown content support and configurable building. My thinking is that the former requires a different model (i.e. indexing) and we haven't made sure whether we broke it at some point in the last while anyway. The latter seemed more useful in the "we're indexing everything" world. For me the added complexity doesn't have benefits that are obvious to me in the reductio world.
Anyway, I took quite the chainsaw swing at the builder, so I think it's quite likely that I missed some nuance here and there and dropped too much. So if you could help me / negotiate with me figure out what we need to bring back, that'd be helpful.

___

That should be the outline of what's left for feature parity. There's also performance parity, on my machine the full build is around 50% slower atm. I don't worry about this too much atm, as I don't think there's any optimization we did before that we can't do now. Just want to wait til I'm sure about the output being correct before I dive more into it.

peterbe commented:
I look forward to learning from this. 
But, quick question; why do you leave out the "whole build-mode complexity"? Without a performant builder does it have any hope at all if we're going to jamstack deploy this?

Gregoor commented:
I've put that in misleading way! By not bother with building I meant just using the builder as is (which requires some reverting of things I threw away and sharing of code between the two builders).

escattone commented on content/utils.js:
```diff
@@ -0,0 +1,53 @@
+
+  const cache = new Map();
+  return (...args) => {
+    const key = JSON.stringify(args);
```
If multiple calls to `fn` include objects in `args` but don't use the same key order for those objects, then the memoization will be replicated. It might not be worth making this more robust in its handling of objects, but just something to keep in mind

Gregoor replied:
Yup, good point! I also considered explicitly having a cache key getter as a parameter, but the call sites were simple enough that it seemed fine. I'll add a comment though, explaining the limitation.

escattone commented on content/document.js:
```diff
@@ -206,16 +193,42 @@ function update(contentRoot, folder, rawHtml, metadata) {
+  updateWikiHistory(path.join(CONTENT_ROOT, metadata.locale), metadata.slug);
+}
+
+const findByURL = memoizeDuringBuild((url, fields = null) => {
```
This function has to be more robust in terms of its handling of the incoming `url`, since the incoming `url` can come from the arguments of macro calls (which are notoriously unreliable). It'll have to do two things:
- repair an unorthodox URL (like we do currently with https://github.com/mdn/yari/blob/674e331d2f79f3c340339f31cf80b387b2bf7e60/content/scripts/build.js#L244)
- replace a redirected URL with its new target URL

so basically it'll have to do what the current [`cleanUri`](https://github.com/mdn/yari/blob/674e331d2f79f3c340339f31cf80b387b2bf7e60/content/scripts/build.js#L369) does.

escattone replied:
Oh, I see you already have this functionality within `info.js` as `return Redirect.resolve(repairURL(uri));` within `getUriKey()`. I wonder if `repairURL(uri)` belongs better in `content/utils.js`?

escattone replied:
Actually, the more I think about this, we only really need to clean the URL's arriving from the content (e.g., arguments to macro calls), so I suppose we could just pre-wrap the `globals.renderFromURL` within `kumascript/src/environment.js`, so something like:
```js
globals.renderFromURL = (url) => renderFromURL(getUriKey(url));
```

escattone commented on kumascript/index.js:
```diff
@@ -1,79 +1,54 @@
+    {
+      ...{
+        path: url,
+        url: `${"this.options.sitemapBaseUrl"}${url}`,
```
I suppose we should create a new constant (maybe just `SITE_URL`?) that can be overridden from the environment for this? Basically it would be the domain we're deploying to (and its protocol), so for example `https://developer.mozilla.org` or `https://pr9999.yari.io` or `http://localhost:5000` or something like that.

Gregoor replied:
Good idea, this was a bit of an easy to overlook placeholder for something that still needs to be worked on, thanks for highlighting it!

Gregoor replied:
Out of curiousity: Why do we want to use an absolute URL here? Couldn't we just as well skip the host?

escattone commented on kumascript/src/api/wiki.js:
```diff
@@ -69,7 +69,7 @@ module.exports = {
+    let result = await this.renderFromURL(path);
     const pathDescription = this.info.getDescription(path);
 
     if (!result) {
```
With your new approach, this entire `if` block is no longer needed.

Gregoor replied:
Awesome, removed! Thanks

escattone commented on kumascript/src/info.js:
```diff
@@ -162,48 +68,81 @@ class AllPagesInfo {
-      // Many of the macros expect an empty object
-      // when the requested document doesn't exist.
-      return {};
+    const result = Document.findByURL(url, { metadata: true });
```
You'll need to clean the incoming `url`, so:
```suggestion
    const result = Document.findByURL(getUriKey(url), { metadata: true });
```

Gregoor replied:
Fixed, thank you!

escattone commented on kumascript/src/info.js:
```diff
@@ -1,149 +1,55 @@
+  },
 
   getDescription(url) {
     const uriKey = this.getUriKey(url);
```
Is it safer to use `info.getUriKey` instead?

Gregoor replied:
Yes, much easier on the eyes. Thanks for the suggestion

escattone commented on kumascript/src/info.js:
```diff
@@ -162,48 +68,81 @@ class AllPagesInfo {
+      tags: tags || [],
+      translations: [], //TODO Object.freeze(buildTranslationObjects(data)),
+      get subpages() {
+        return Document.findChildren(url, { metadata: true }).map((document) =>
```
Probably better to use `document.url` here since it has already been "cleaned":
```suggestion
        return Document.findChildren(document.url, { metadata: true }).map((document) =>
```

escattone commented on content/document.js:
```diff
@@ -206,16 +193,42 @@ function update(contentRoot, folder, rawHtml, metadata) {
+  const folder = urlToFolderPath(url);
+  const childPaths = glob.sync(
+    path.join(CONTENT_ROOT, folder, "*", HTML_FILENAME),
+    {
+      ignore: path.join(CONTENT_ROOT, getHTMLPath(folder)),
+    }
```
I don't think this is needed, at least it worked for me without it.

Gregoor replied:
Indeed!

escattone commented on kumascript/src/environment.js:
```diff
@@ -133,6 +138,7 @@ class Environment {
     globals.page = globals.Page = freeze(page);
     globals.env = globals.Env = freeze(env);
     globals.info = freeze(info);
+    globals.renderFromURL = renderFromURL;
```
Since this function will receive URL's from the "wilds" of the content, it's important to clean the incoming URL's and maybe it's best to do that here?
```suggestion
    globals.renderFromURL = (url) => renderFromURL(info.getUriKey(url));
```

Gregoor commented:
Alrighty, you two. The framework is in place. Here's my idea for the way forward (ofc feel free to make other suggestions):

I'd work on (re)adding
- [ ] sitemaps
- [ ] make the functional tests pass
- [ ] CRUD
- [ ] Document watcher

___

@peterbe & @escattone: These are some other bigger remaining TODOs, I don't have a clear grasp on who'd be best for which one. But if you want to tell me which one(s) you want to take on, I'd create some sprint task issues to satisfy the sprint gods.

## translations
As we three talked about the current lookup model doesn't work for reductio (`translation_of` living in the translated document, hence an O(1) lookup not being possible). It sounded like we were all vaguely in favor of having something like a `translations.json` in the locale root (just like `wiki_history.json`). But I don't want to speak for anyone else if there's still disagreement here.
Anyway, if there isn't disagreement. This would require a change to the importer (which now lives in its own workspace)

## flaws
I think I broke some things there and I haven't backported (err frontported?) fixable flaws yet. Also happy to take this on as I broke it, but if you feel inclined to, that's also great.

## everything KS
I touched a lot of that code and haven't exhaustively checked the results yet. After a general review is done, I'd wanna compare output of the builder on `master` with this branch's output.

## things that were lost
Not on my todo list are stumptown content support and configurable building. My thinking is that the former requires a different model (i.e. indexing) and we haven't made sure whether we broke it at some point in the last while anyway. The latter seemed more useful in the "we're indexing everything" world. For me the added complexity doesn't have benefits that are obvious to me in the reductio world.
Anyway, I took quite the chainsaw swing at the builder, so I think it's quite likely that I missed some nuance here and there and dropped too much. So if you could help me / negotiate with me figure out what we need to bring back, that'd be helpful.

___

That should be the outline of what's left for feature parity. There's also performance parity, on my machine the full build is around 2x slower atm. I don't worry about this too much atm, as I don't think there's any optimization we did before that we can't do now. Just want to wait til I'm sure about the output being correct before I dive more into it.

Gregoor commented:
Oh updating the docs is another thing I want to do, which will mainly be subtractive work. All the commands are still the same, there'd probably just be less of them.

Gregoor commented:
Okay things are a little intertwined: to fix the remaining tests, I need to fix flaws first 🙃 

Also had the audacity to comment out some tests that checked for `modified` and `popularity` being present. I'm wondering whether we'd want to change those tests to check for surface area instead of internal data structures anyway? `modified` is not used anywhere yet though, with popularity we could check for the order of the suggested items (though to a degree we're also at the mercy of our search indexing library there). Wdyt @peterbe?

peterbe commented:
> Also had the audacity to comment out some tests that checked for `modified` and `popularity` being present. I'm wondering whether we'd want to change those tests to check for surface area instead of internal data structures anyway? `modified` is not used anywhere yet though, with popularity we could check for the order of the suggested items (though to a degree we're also at the mercy of our search indexing library there). Wdyt @peterbe?

Quick drive-by comment; The `modified` is used in the `kumastyles` branch. At some point, we should check that presence with headless tests, like a human QA would do. 
Also, the popularity *number* is never important. So yes, we can totally look at the order they appear. 
Headless tests are always better than "jest tests" but some things can weird/tricky to test with headless tests. But what always matters the most is that the functionality is there.

escattone commented:
@Gregoor @peterbe I'll start by taking `Everything KS` and the KS-related flaws if that's cool with you guys.

peterbe commented:
Clicking "Edit in your editor" fails with:
<img width="747" alt="Screen Shot 2020-07-21 at 12 27 40 PM" src="https://user-images.githubusercontent.com/26739/88080832-991daa80-cb4d-11ea-9c68-d7411fdddd97.png">

The in-browser editor fails with:
```
12:28:36 PM server.1   |  (node:81307) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'metadata' of null
12:28:36 PM server.1   |      at Object.update (/Users/peterbe/dev/MOZILLA/MDN/yari/content/document.js:150:28)
12:28:36 PM server.1   |      at /Users/peterbe/dev/MOZILLA/MDN/yari/server/document.js:34:14
12:28:36 PM server.1   |      at Layer.handle [as handle_request] (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/layer.js:95:5)
12:28:36 PM server.1   |      at next (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/route.js:137:13)
12:28:36 PM server.1   |      at withDocument (/Users/peterbe/dev/MOZILLA/MDN/yari/server/document.js:24:3)
12:28:36 PM server.1   |      at Layer.handle [as handle_request] (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/layer.js:95:5)
12:28:36 PM server.1   |      at next (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/route.js:137:13)
12:28:36 PM server.1   |      at Route.dispatch (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/route.js:112:3)
12:28:36 PM server.1   |      at Layer.handle [as handle_request] (/Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/layer.js:95:5)
12:28:36 PM server.1   |      at /Users/peterbe/dev/MOZILLA/MDN/yari/node_modules/express/lib/router/index.js:281:22
12:28:36 PM server.1   |  (node:81307) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 8)
```

Gregoor commented:
Thanks Peter I fixed this in one go with CRUD.

Also I'm treating [my comment from above](https://github.com/mdn/yari/pull/881#issue-447610018) as the TODO list for this to land (and I moved it into the first post). There's an udpate to translations in there from Ryan and I talking about it yesterday.

Gregoor commented:
Thanks for all the effort getting this to parity! We're close. Today I've put live samples and web-socket based change notifications in. For both I'd appreciate a check whether it generally makes sense. For the former probably from @escattone, for the latter from @peterbe.

Also I could use some guidance wrt merging master into this branch. Anything I should look out for when merging kumascript?

escattone commented:
@Gregoor Awesome, I'll check out the live samples. Regarding guidance on how to merge master into this branch, for the `kumascript` workspace I would say that we always take `reductio-reloaded` rather than what's in`master`, but I'll double check that.

Gregoor commented:
I'm a bit confused, the headless and search tests run fine for me. A complete test run yields success for 13/16. Whereas in GH action we only have 8/16. They all seem to be related to timeouts which I don't have an explanation for yet.

escattone commented:
@Gregoor I think I found the problem. The `buildLiveSamplePageFromURL()` function within the preview server always gets lower-cased URL's, so when it extracts the live-sample ID from the URL, that's always lower-case as well. However, the sample D's are case sensitive, so the live-samples with ID's containing upper-case letters are never found. More tommorrow.

escattone commented:
@Gregoor The fix is https://github.com/Gregoor/yari/pull/13.
